### DISCUSSION TITLE
Send action back into the system when effect is canceled

### DISCUSSION THREAD

**Discussion by vykut** [pointfreeco/swift-composable-architecture#2670] [Q&A] [answered]
Hi,

In my scenario I have a long running effect that makes some requests to the local database as well as a network API in order to fetch some data. 
Because this effect emits from time to time via `send(.delegate(.searchResults(...)))`, I store a boolean in the State e.g `isSearching`. 
In case the effect gets canceled I would like to be able to flip the `isSearching` boolean back to false, but the `send` callback stops working once the task has been canceled. 

Any ideas how I can/ should model my logic?

**Comment by vykut** [comment]
Hi @mbrandonw and @stephencelis, I would greatly appreciate your inputs on this 🙏

**Reply by mbrandonw** [reply, marked as answer]
You can use `.concatenate` to send an action when an effect completes or is cancelled:

```swift
return .concatenate(
  .run { send in 
    …
  }
  .cancellable(id: …),

  .send(.searchFinished)
)
```

**Reply by vykut** [reply]
I just stumbled upon `withTaskCancellation` and I see it provides similar functionality.
Refactoring your example to use `withTaskCancellation`, yields the same results, unless I'm missing something...
```swift
return .run { send in
    await send(.searchStarted)
    try? await withTaskCancellation(id: CancelID.search, cancelInFlight: true) {
        try await self.clock.sleep(for: .milliseconds(300))
        // network request
        await send(.delegate(.result(...)))
        // other network request
        await send(.delegate(.result(...)))
    }
    await send(.searchFinished)
}
```

**Reply by mbrandonw** [reply]
Hi @vykut yeah that will work too.
